r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,1015 @@
+                    // Returns bits in big-endian order
+                    for (i, b) in BitIterator::new(r_minus_1).enumerate() {
+                        let mut cs = cs.ns(|| format!("Iteration {}", i));
+                        let old_seen_one = seen_one;
+                        if seen_one {
+                            result.double_in_place(cs.ns(|| "Double"))?;
+                        } else {
+                            seen_one = b;
+                        }
+                        if b {
+                            result = if old_seen_one {
+                                result.add(cs.ns(|| "Add"), &ge)?
+                            } else {
+                                ge.clone()
+                            };
+                        }
+                    }
+                    let neg_ge = ge.negate(cs.ns(|| "Negate ge"))?;
+                    neg_ge.enforce_equal(cs.ns(|| "Check equals"), &result)?;
+                    Ok(ge)
+                }
+            };
+        let ge = alloc_and_prime_order_check(
+            cs.ns(|| "alloc and prime order check"),
+            value_gen
+        )?;
+
+        Ok(ge)
+    }
+
+    #[inline]
+    fn alloc_input<FN, T, CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        value_gen: FN,
+    ) -> Result<Self, SynthesisError>
+        where
+            FN: FnOnce() -> Result<T, SynthesisError>,
+            T: Borrow<SWProjective<P>>,
+    {
+        let (x, y, infinity) = match value_gen() {
+            Ok(ge) => {
+                let ge = ge.borrow().into_affine();
+                (Ok(ge.x), Ok(ge.y), Ok(ge.infinity))
+            },
+            _ => (
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+            ),
+        };
+
+        let b = P::COEFF_B;
+        let a = P::COEFF_A;
+
+        let x = NonNativeFieldGadget::alloc_input(&mut cs.ns(|| "x"), || x)?;
+        let y = NonNativeFieldGadget::alloc_input(&mut cs.ns(|| "y"), || y)?;
+        let infinity = Boolean::alloc_input(&mut cs.ns(|| "infinity"), || infinity)?;
+
+        // Check that y^2 = x^3 + ax +b
+        // We do this by checking that y^2 - b = x * (x^2 +a)
+        // let x2 = x.square(&mut cs.ns(|| "x^2"))?;
+        // let y2 = y.square(&mut cs.ns(|| "y^2"))?;
+        let x2 = x.mul(cs.ns(|| "x^2"), &x)?;
+        let y2 = y.mul(cs.ns(|| "y^2"), &y)?;
+
+
+        let x2_plus_a = x2.add_constant(cs.ns(|| "x^2 + a"), &a)?;
+        let y2_minus_b = y2.add_constant(cs.ns(|| "y^2 - b"), &b.neg())?;
+
+        let x2_plus_a_times_x = x2_plus_a.mul(cs.ns(|| "(x^2 + a)*x"), &x)?;
+
+        x2_plus_a_times_x.conditional_enforce_equal(
+            cs.ns(|| "on curve check"),
+            &y2_minus_b,
+            &infinity.not()
+        )?;
+
+        Ok(Self::new(x, y, infinity))
+    }
+}
+
+
```

Comment:
We got tests for all the GroupGadgets implementation. Can you add for each of them test cases also for NonNativeGroupGadget and check that everything works properly ?

Reply:
Tests added. Constraints for exponentiation are too much though. Investigating.